src/main/java/com/powsybl/caseserver/service/S3CaseService.java

```diff
@@ -0,0 +1,558 @@
+    }
+
+    @Override
+    public UUID importCase(MultipartFile mpf, boolean withExpiration, boolean withIndexation) {
+        UUID caseUuid = UUID.randomUUID();
+
+        String caseName = Objects.requireNonNull(mpf.getOriginalFilename());
+        validateCaseName(caseName);
+
+        String format = withTempCopy(caseUuid, caseName, mpf::transferTo, this::getFormat);
+        String compressionFormat = null;
+
+        try (InputStream inputStream = mpf.getInputStream()) {
+            String key = uuidAndFilenameToKey(caseUuid, caseName);
+
+            Map<String, String> caseMetadata = new HashMap<>();
+            caseMetadata.put(FORMAT_HEADER_KEY, format);
+
+            if (isArchivedCaseFile(caseName)) {
+                importZipContent(mpf.getInputStream(), caseUuid);
+                compressionFormat = FileNameUtils.getExtension(Paths.get(caseName));
+            } else if (isCompressedCaseFile(caseName)) {
+                compressionFormat = FileNameUtils.getExtension(Paths.get(caseName));
+            }
+
+            PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                    .bucket(bucketName)
+                    .key(key)
+                    .metadata(caseMetadata)
+                    .contentType(mpf.getContentType())
+                    .build();
+
+            // Use putObject to upload the file
+            s3Client.putObject(putObjectRequest, RequestBody.fromInputStream(inputStream, mpf.getSize()));
+
+        } catch (IOException e) {
+            throw CaseException.createFileNotImportable(caseName, e);
+        }
+
+        createCaseMetadataEntity(caseUuid, withExpiration, withIndexation, caseName, compressionFormat, caseMetadataRepository);
+        CaseInfos caseInfos = createInfos(caseName, caseUuid, format);
+        if (withIndexation) {
+            caseInfosService.addCaseInfos(caseInfos);
+        }
+        notificationService.sendImportMessage(caseInfos.createMessage());
+
+        return caseUuid;
+    }
+
+    private void importZipContent(InputStream inputStream, UUID caseUuid) throws IOException {
+        try (ZipInputStream zipInputStream = new SecuredZipInputStream(inputStream, 1000, MAX_SIZE)) {
+            ZipEntry entry;
+
+            while ((entry = zipInputStream.getNextEntry()) != null) {
+                if (!entry.isDirectory()) {
+                    processEntry(caseUuid, zipInputStream, entry);
+                }
+                zipInputStream.closeEntry();
+            }
+        }
+    }
+
+    private void processEntry(UUID caseUuid, ZipInputStream zipInputStream, ZipEntry entry) throws IOException {
+        String fileName = entry.getName();
+        String extractedKey = uuidAndFilenameToKey(caseUuid, fileName);
+
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        byte[] buffer = new byte[1024];
+        int length;
+        while ((length = zipInputStream.read(buffer)) > 0) {
```

Comment:
I think we have a method to copy a stream to another in guava or commons I don't remember we already use it

Reply:
Yes :
byte[] fileBytes = ByteStreams.toByteArray(zipInputStream);